File: lib/api/twitch_api_cache.dart
import "dart:async";
import "dart:convert";

import "package:flow/api/twitch_api.dart";

typedef TwitchApiClientLoader = Future<TwitchApiClient> Function();

class TwitchApiCache {
  TwitchApiCache({
    required this.clientLoader,
    this.maxEntries = 128,
  }) : assert(maxEntries > 0);

  final TwitchApiClientLoader clientLoader;
  final int maxEntries;
  final _values = <String, Object?>{};
  final _inFlight = <String, Future<Object?>>{};
  void clear() {
    _values.clear();
    _inFlight.clear();
  }

  Future<TwitchPage<TwitchCategory>> fetchTopCategoriesPage({
    int first = 12,
    String? cursor,
    bool refresh = false,
  }) => _cached(
    _cacheKey("topCategories", {
      "first": first,
      "cursor": cursor,
    }),
    (client) => client.fetchTopCategoriesPage(first: first, cursor: cursor),
    refresh: refresh,
  );

  Future<TwitchPage<TwitchFollowedStream>> fetchLiveStreamsPage({
    int first = 20,
    List<String> gameIds = const <String>[],
    List<String> userLogins = const <String>[],
    String? cursor,
    bool refresh = false,
  }) => _cached(
    _cacheKey("liveStreams", {
      "first": first,
      "gameIds": gameIds,
      "userLogins": _normalizedValues(userLogins),
      "cursor": cursor,
    }),
    (client) => client.fetchLiveStreamsPage(
      first: first,
      gameIds: gameIds,
      userLogins: userLogins,
      cursor: cursor,
    ),
    refresh: refresh,
  );

  Future<Map<String, TwitchUser>> fetchUsersByIds(
    List<String> ids, {
    bool refresh = false,
  }) => _cached(
    _cacheKey("usersByIds", {"ids": _normalizedValues(ids)}),
    (client) => client.fetchUsersByIds(ids),
    refresh: refresh,
  );

  Future<TwitchPage<TwitchSearchChannel>> searchChannelsPage(
    String query, {
    int first = 20,
    String? cursor,
    bool liveOnly = false,
    bool refresh = false,
  }) => _cached(
    _cacheKey("searchChannels", {
      "query": query.trim().toLowerCase(),
      "first": first,
      "cursor": cursor,
      "liveOnly": liveOnly,
    }),
    (client) => client.searchChannelsPage(
      query,
      first: first,
      cursor: cursor,
      liveOnly: liveOnly,
    ),
    refresh: refresh,
  );

  Future<TwitchChannelDetails> fetchChannelDetails(
    String login, {
    int videosFirst = 30,
    String? videosCursor,
    bool refresh = false,
  }) => _cached(
    _cacheKey("channelDetails", {
      "login": login.trim().toLowerCase(),
      "videosFirst": videosFirst,
      "videosCursor": videosCursor,
    }),
    (client) => client.fetchChannelDetails(
      login,
      videosFirst: videosFirst,
      videosCursor: videosCursor,
    ),
    refresh: refresh,
  );

  Future<Uri> fetchLivePlaybackUri(String login) async {
    final client = await clientLoader();
    return client.fetchLivePlaybackUri(login);
  }

  Future<bool> fetchChannelSubscriptionStatus(String login) async {
    final client = await clientLoader();
    return client.fetchChannelSubscriptionStatus(login);
  }

  Future<TwitchPage<TwitchCategory>> searchCategoriesPage(
    String query, {
    int first = 20,
    String? cursor,
    bool refresh = false,
  }) => _cached(
    _cacheKey("searchCategories", {
      "query": query.trim().toLowerCase(),
      "first": first,
      "cursor": cursor,
    }),
    (client) => client.searchCategoriesPage(query, first: first, cursor: cursor),
    refresh: refresh,
  );

  Future<T> _cached<T>(
    String key,
    Future<T> Function(TwitchApiClient client) load, {
    required bool refresh,
  }) async {
    if (!refresh && _values.containsKey(key)) {
      final value = _values.remove(key);
      _values[key] = value;
      return value as T;
    }

    if (!refresh) {
      final pending = _inFlight[key];
      if (pending != null) {
        return (await pending) as T;
      }
    }

    final future = _loadValue(load);
    _inFlight[key] = future;

    try {
      final value = await future;
      if (identical(_inFlight[key], future)) {
        _values.remove(key);
        _values[key] = value;
        while (_values.length > maxEntries) {
          _values.remove(_values.keys.first);
        }
      }
      return value;
    } finally {
      if (identical(_inFlight[key], future)) {
        unawaited(_inFlight.remove(key));
      }
    }
  }

  Future<T> _loadValue<T>(
    Future<T> Function(TwitchApiClient client) load,
  ) async => load(await clientLoader());
}

String _cacheKey(String namespace, Map<String, Object?> values) => jsonEncode([namespace, values]);

List<String> _normalizedValues(Iterable<String> values) => {
  for (final value in values)
    if (value.trim().isNotEmpty) value.trim(),
}.toList()..sort();

File: android/app/src/main/kotlin/com/namecallfilter/flow/data/TwitchApiCache.kt
package com.namecallfilter.flow.data

import kotlinx.coroutines.CompletableDeferred
import java.net.URI
import java.util.Locale

/** Small session-scoped LRU with request coalescing; playback/subscription are intentionally uncached. */
class TwitchApiCache(
    private val clientLoader: suspend () -> TwitchApi,
    val maxEntries: Int = 128,
) {
    init {
        require(maxEntries > 0)
    }

    private val lock = Any()
    private val values = LinkedHashMap<String, Any?>(16, 0.75f, true)
    private val inFlight = mutableMapOf<String, CompletableDeferred<Any?>>()
    private var revision = 0

    fun clear() {
        synchronized(lock) {
            revision += 1
            values.clear()
            inFlight.clear()
        }
    }

    suspend fun fetchTopCategoriesPage(
        first: Int = 12,
        cursor: String? = null,
        refresh: Boolean = false,
    ): TwitchPage<TwitchCategory> = cached(
        cacheKey("topCategories", mapOf("first" to first, "cursor" to cursor)),
        refresh,
    ) { it.fetchTopCategoriesPage(first, cursor) }

    suspend fun fetchLiveStreamsPage(
        first: Int = 20,
        gameIds: List<String> = emptyList(),
        userLogins: List<String> = emptyList(),
        cursor: String? = null,
        refresh: Boolean = false,
    ): TwitchPage<TwitchFollowedStream> = cached(
        cacheKey(
            "liveStreams",
            mapOf(
                "first" to first,
                "gameIds" to normalizedValues(gameIds),
                "userLogins" to normalizedValues(userLogins),
                "cursor" to cursor,
            ),
        ),
        refresh,
    ) { it.fetchLiveStreamsPage(first, gameIds, userLogins, cursor) }

    suspend fun fetchUsersByIds(
        ids: List<String>,
        refresh: Boolean = false,
    ): Map<String, TwitchUser> = cached(
        cacheKey("usersByIds", mapOf("ids" to normalizedValues(ids))),
        refresh,
    ) { it.fetchUsersByIds(ids) }

    suspend fun searchChannelsPage(
        query: String,
        first: Int = 20,
        cursor: String? = null,
        liveOnly: Boolean = false,
        refresh: Boolean = false,
    ): TwitchPage<TwitchSearchChannel> = cached(
        cacheKey(
            "searchChannels",
            mapOf(
                "query" to query.trim().lowercase(Locale.ROOT),
                "first" to first,
                "cursor" to cursor,
                "liveOnly" to liveOnly,
            ),
        ),
        refresh,
    ) { it.searchChannelsPage(query, first, cursor, liveOnly) }

    suspend fun fetchChannelDetails(
        login: String,
        videosFirst: Int = 30,
        videosCursor: String? = null,
        refresh: Boolean = false,
    ): TwitchChannelDetails = cached(
        cacheKey(
            "channelDetails",
            mapOf(
                "login" to login.trim().lowercase(Locale.ROOT),
                "videosFirst" to videosFirst,
                "videosCursor" to videosCursor,
            ),
        ),
        refresh,
    ) { it.fetchChannelDetails(login, videosFirst, videosCursor) }

    suspend fun searchCategoriesPage(
        query: String,
        first: Int = 20,
        cursor: String? = null,
        refresh: Boolean = false,
    ): TwitchPage<TwitchCategory> = cached(
        cacheKey(
            "searchCategories",
            mapOf(
                "query" to query.trim().lowercase(Locale.ROOT),
                "first" to first,
                "cursor" to cursor,
            ),
        ),
        refresh,
    ) { it.searchCategoriesPage(query, first, cursor) }

    suspend fun fetchLivePlaybackUri(login: String): URI =
        clientLoader().fetchLivePlaybackUri(login)

    suspend fun fetchChannelSubscriptionStatus(login: String): Boolean =
        clientLoader().fetchChannelSubscriptionStatus(login)

    @Suppress("UNCHECKED_CAST")
    private suspend fun <T> cached(
        key: String,
        refresh: Boolean,
        load: suspend (TwitchApi) -> T,
    ): T {
        var leader = false
        var loadRevision = 0
        val deferred = synchronized(lock) {
            if (!refresh && values.containsKey(key)) {
                return values[key] as T
            }
            if (!refresh) inFlight[key]?.let { return@synchronized it }

            leader = true
            loadRevision = revision
            CompletableDeferred<Any?>().also { inFlight[key] = it }
        }
        if (!leader) return deferred.await() as T

        try {
            val value = load(clientLoader())
            deferred.complete(value)
            synchronized(lock) {
                if (loadRevision == revision) {
                    values.remove(key)
                    values[key] = value
                    while (values.size > maxEntries) {
                        values.remove(values.entries.first().key)
                    }
                }
            }
            return value
        } catch (error: Throwable) {
            deferred.completeExceptionally(error)
            throw error
        } finally {
            synchronized(lock) {
                if (loadRevision == revision && inFlight[key] === deferred) inFlight.remove(key)
            }
        }
    }
}

internal fun cacheKey(namespace: String, values: Map<String, Any?>): String = buildList {
    add(namespace)
    values.toSortedMap().forEach { (key, value) -> add("$key=${cacheValue(value)}") }
}.joinToString("|")

private fun cacheValue(value: Any?): String = when (value) {
    null -> ""
    is Iterable<*> -> normalizedValues(value.filterIsInstance<String>()).joinToString(",")
    else -> value.toString()
}

private fun normalizedValues(values: Iterable<String>): List<String> = values
    .map(String::trim)
    .filter(String::isNotEmpty)
    .sorted()
